apps/api/src/tasks/pending-tasks.service.ts

```diff
@@ -43,110 +56,188 @@ export class PendingTasksService {
+   *   @param params.taskStatus - The status of the task.
+   *   @param params.taskName - Optional. The name of the task.
+   *   @param params.exclude - Optional. An array of task IDs to exclude from the search.
+   *
+   *   @returns An object containing the options for the find operation.
+   */
+  private getOptions(params: {
+    runId: number;
+    priority: TaskPriority;
+    taskStatus: TaskStatus;
+    taskName?: string;
+    exclude?: number[];
+  }): FindOneOptions<TaskEntity> {
+    const { runId, priority, taskStatus, taskName, exclude = [] } = params;
+
+    // Set the order based on the priority
+    let order;
+    switch (priority) {
+      case TaskPriority.Longest:
+        order = { avgDuration: 'DESC' };
+        break;
+      case TaskPriority.Shortest:
+        order = { avgDuration: 'ASC' };
+        break;
+      case TaskPriority.Newest:
+        order = { createdAt: 'DESC' }; // FIFO
+        break;
+      case TaskPriority.Oldest:
+        order = { createdAt: 'ASC' }; // LIFO
+    }
 
-  private getShortestOptions(runId: number): FindOneOptions<TaskEntity> {
     return {
       where: {
         run: { id: runId },
-        status: TaskStatus.Pending,
-        priority: TaskPriority.Shortest,
+        name: taskName,
+        status: taskStatus,
+        priority: priority,
+        id: Not(In(exclude)),
       },
-      order: { avgDuration: 'ASC' },
+      order: order,
       lock: {
         mode: 'pessimistic_write',
         tables: ['task'],
       },
     };
   }
 
-  // FIFO
-  private getNewestOptions(runId: number): FindOneOptions<TaskEntity> {
-    return {
-      where: {
-        run: { id: runId },
-        status: TaskStatus.Pending,
-        priority: TaskPriority.Newest,
-      },
-      order: { createdAt: 'DESC' },
-      lock: {
-        mode: 'pessimistic_write',
-        tables: ['task'],
-      },
-    };
-  }
+  /**
+   * This function is used to find a task based on the provided options.
+   * It checks for circular dependencies, whether all dependencies of a task are completed,
+   * and if not, it finds the next task that can be executed.
+   *
+   * @param params - An object containing the parameters for the find operation.
+   * @param params.manager - The EntityManager instance.
+   * @param params.runId - The ID of the run.
+   * @param params.priority - The priority of the task.
+   * @param params.task - Optional. The task entity.
+   * @param params.seenTasks - Optional. A Map of seen tasks to avoid circular dependencies.
+   * @param params.searchForNext - Optional. A boolean to decide whether to search for the next task if the current task is blocked.
+   *
+   * @returns The task that can be executed next, or null if no such task is found.
+   */
+  private async findTask(params: {
+    manager: EntityManager;
+    runId: number;
+    priority: TaskPriority;
+    task?: TaskEntity;
+    seenTasks?: Map<number, string>;
+    searchForNext?: boolean;
+  }): Promise<TaskEntity> {
```

Comment:
I think we need a method, findAllPendingTasks and then execute the rest of the logic from those tasks. Did you try this approach instead of doing it one by one ?